### ISSUE TITLE
Ubuntu PPA package for 16.04+ does not provide override for the /usr/lib/tmpfiles.d/var.conf file

### ISSUE THREAD

**Issue by deoren** [rsyslog/rsyslog#1655]
## Scenario where this matters:

1. Ubuntu 14.04 is installed, stock Ubuntu repo-provided rsyslog is working normally, able to dynamically generate log files within `/var/log` .
1. User enables PPA and installs rsyslog. rsyslog works normally.
1. User upgrades to 16.04. 
1. The `/usr/lib/tmpfiles.d/var.conf` config file specifies permissions at boot as 0755 on `/var/log`, the owner is `root:syslog` and rsyslog is configured to drop privileges to `syslog:syslog`.
1. The dynamic creation of new log files within `/var/log` is now broken because there is not another config file set to override the values set within `/usr/lib/tmpfiles.d/var.conf`

## Expected behavior:

The rsyslog package provided by the Ubuntu repos overrides the settings within `/usr/lib/tmpfiles.d/var.conf` to allow the privilege drop functionality to work as expected, giving rsyslog write permission (via group membership) directly to `/var/log` so that files can be dynamically created as necessary.

## Suggested fix:

Update the PPA package to install an override conf file (e.g., `/usr/lib/tmpfiles.d/00rsyslog.conf`) to mirror what Ubuntu is already doing with their rsyslog package. This should handle use cases where someone installs the package via PPA when running Ubuntu 14.04 and then upgrading their release version sometime after that.

**Comment by deoren** [comment]
I started to create a separate issue for this, but it appears that since the `/usr/lib/tmpfiles.d/00rsyslog.conf` file is not considered a configuration file (missing from `conffiles` within the package), when you upgrade from the Ubuntu-provided rsyslog package to the PPA-provided package it is removed.

This breaks the expected behavior of allowing dynamic log file creation within `/var/log`.

## Steps to reproduce

1. Confirm that the `/usr/lib/tmpfiles.d/00rsyslog.conf` file is actually there
1. Enable the PPA
1. Update apt's package cache
1. Upgrade rsyslog
1. Confirm that the `/usr/lib/tmpfiles.d/00rsyslog.conf` file is no longer present

## Details

Below is a test run of enabling the PPA and upgrading the existing rsyslog installation.

`root@ubuntu-virtual-machine:~# ls -l /{etc,usr/lib}/tmpfiles.d/*rsys*`
<pre>
ls: cannot access '/etc/tmpfiles.d/*rsys*': No such file or directory
-rw-r--r-- 1 root root 238 Dec  2  2015 /usr/lib/tmpfiles.d/00rsyslog.conf
</pre>

`root@ubuntu-virtual-machine:~# apt-add-repository ppa:adiscon/v8-stable`
<pre>
 Contains the latest RSyslog V8-Stable packages and dependencies.
 More info: https://launchpad.net/~adiscon/+archive/ubuntu/v8-stable
Press [ENTER] to continue or ctrl-c to cancel adding it

gpg: keyring `/tmp/tmpz21fr26a/secring.gpg' created
gpg: keyring `/tmp/tmpz21fr26a/pubring.gpg' created
gpg: requesting key 5234BF2B from hkp server keyserver.ubuntu.com
gpg: /tmp/tmpz21fr26a/trustdb.gpg: trustdb created
gpg: key 5234BF2B: public key "Launchpad PPA for Adiscon" imported
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
OK
</pre>

`root@ubuntu-virtual-machine:~# apt-get update > /dev/null`
`root@ubuntu-virtual-machine:~# apt-get install rsyslog`
<pre>
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  linux-headers-4.4.0-79 linux-headers-4.4.0-79-generic linux-image-4.4.0-79-generic linux-image-extra-4.4.0-79-generic
Use 'sudo apt autoremove' to remove them.
The following additional packages will be installed:
  libfastjson4 liblogging-stdlog1
Suggested packages:
  rsyslog-mysql | rsyslog-pgsql rsyslog-doc rsyslog-gssapi rsyslog-relp rsyslog-elasticsearch rsyslog-mmjsonparse rsyslog-imptcp rsyslog-gnutls rsyslog-udpspoof
  rsyslog-mmrm1stspace rsyslog-kafka rsyslog-pmnull
The following NEW packages will be installed:
  libfastjson4 liblogging-stdlog1
The following packages will be upgraded:
  rsyslog
1 upgraded, 2 newly installed, 0 to remove and 39 not upgraded.
Need to get 633 kB of archives.
After this operation, 708 kB of additional disk space will be used.
Do you want to continue? [Y/n] y
Get:1 http://ppa.launchpad.net/adiscon/v8-stable/ubuntu xenial/main amd64 libfastjson4 amd64 0.99.5-0adiscon1xenial1 [18.7 kB]
Get:2 http://ppa.launchpad.net/adiscon/v8-stable/ubuntu xenial/main amd64 liblogging-stdlog1 amd64 1.0.6-0adiscon1xenial1 [2,480 B]
Get:3 http://ppa.launchpad.net/adiscon/v8-stable/ubuntu xenial/main amd64 rsyslog amd64 8.27.0-0adiscon4xenial1 [612 kB]
Fetched 633 kB in 1s (364 kB/s)  
Selecting previously unselected package libfastjson4.
(Reading database ... 321550 files and directories currently installed.)
Preparing to unpack .../libfastjson4_0.99.5-0adiscon1xenial1_amd64.deb ...
Unpacking libfastjson4 (0.99.5-0adiscon1xenial1) ...
Selecting previously unselected package liblogging-stdlog1.
Preparing to unpack .../liblogging-stdlog1_1.0.6-0adiscon1xenial1_amd64.deb ...
Unpacking liblogging-stdlog1 (1.0.6-0adiscon1xenial1) ...
Preparing to unpack .../rsyslog_8.27.0-0adiscon4xenial1_amd64.deb ...
Unpacking rsyslog (8.27.0-0adiscon4xenial1) over (8.16.0-1ubuntu3) ...
Processing triggers for libc-bin (2.23-0ubuntu9) ...
Processing triggers for ureadahead (0.100.0-19) ...
Processing triggers for systemd (229-4ubuntu17) ...
Processing triggers for man-db (2.7.5-1) ...
Setting up libfastjson4 (0.99.5-0adiscon1xenial1) ...
Setting up liblogging-stdlog1 (1.0.6-0adiscon1xenial1) ...
Setting up rsyslog (8.27.0-0adiscon4xenial1) ...
Installing new version of config file /etc/apparmor.d/usr.sbin.rsyslogd ...
Installing new version of config file /etc/default/rsyslog ...
Installing new version of config file /etc/init.d/rsyslog ...
Installing new version of config file /etc/init/rsyslog.conf ...
Installing new version of config file /etc/logcheck/ignore.d.server/rsyslog ...
Installing new version of config file /etc/logrotate.d/rsyslog ...
Installing new version of config file /etc/rsyslog.conf ...
Replacing config file /etc/rsyslog.d/50-default.conf with new version
Skipping profile in /etc/apparmor.d/disable: usr.sbin.rsyslogd
Processing triggers for libc-bin (2.23-0ubuntu9) ...
Processing triggers for systemd (229-4ubuntu17) ...
Processing triggers for ureadahead (0.100.0-19) ...
</pre>

`root@ubuntu-virtual-machine:~# ls -l /{etc,usr/lib}/tmpfiles.d/*rsys*`
<pre>
ls: cannot access '/etc/tmpfiles.d/*rsys*': No such file or directory
ls: cannot access '/usr/lib/tmpfiles.d/*rsys*': No such file or directory
</pre>

**Comment by davidelang** [comment]
As far as I can tell, /usr/lib/tempfiles.d only exists in the Ubuntu packaging, 
it's not in the adiscon packages or in the Debian packages, so it appears that 
you will need to file an issue with Ubuntu.

**Comment by deoren** [comment]
@davidelang I didn't see `/usr/lib/tempfiles.d` in the PPA package either. Since the PPA package is replacing the Ubuntu package, shouldn't it provide an override of some kind to allow for the same functionality as provided by the Ubuntu package? That way if someone is already relying on that behavior for Ubuntu and they opt to upgrade to the latest upstream package (via PPA) the functionality will remain the same?

**Comment by davidelang** [comment]
if Ubuntu changes what we provide, why should we have to change what we provide? 
How should we find out when the Ubuntu maintainer makes some new change? I have 
no idea when this directory was created in the Ubuntu package.

The problem isn't that we don't include that stuff in conffiles, it's the Ubuntu 
package that should have that defined in conffiles.

There's nothing we can do to prevent the uninstall of the Ubuntu package from 
removing those files, that has to be done in the Ubuntu package definition.

David Lang

**Comment by deoren** [comment]
> How should we find out when the Ubuntu maintainer makes some new change? I have
no idea when this directory was created in the Ubuntu package.

I'm not sure how that would be best handled.

> The problem isn't that we don't include that stuff in conffiles, it's the Ubuntu
package that should have that defined in conffiles.

I understand that argument. I don't know enough about systemd to say whether the `/usr/lib/tmpfiles.d/20rsyslog.conf` file is a core piece of the package or a config file, but since it's not intended to be modify directly by the user (admins should place overrides within `/etc/tmpfiles.d/`) they're probably comfortable with it being removed after the package is removed.

> There's nothing we can do to prevent the uninstall of the Ubuntu package from
removing those files, that has to be done in the Ubuntu package definition.

I understand. I was speaking more of meeting the same functionality level that the Ubuntu package provides. Their package allows for dynamic log file creation within the `/var/log` directory by providing an override (`/usr/lib/tmpfiles.d/00rsyslog.conf``) to the standard systemd configuration file (https://github.com/systemd/systemd/blob/v229/tmpfiles.d/var.conf).

From what I can tell that functionality was introduced in their package [in bug 1539483:](https://launchpad.net/bugs/1539483). I found this in their changelog:

<pre>
rsyslog (8.16.0-1ubuntu1) xenial; urgency=low

  * Merge from Debian unstable (LP: #1539483).  Remaining changes:
    - debian/00rsyslog.conf Install tmpfiles.d snippet to ensure that the syslog
      group can write into /var/log/.
</pre>

I'm checking Debian 8 and 9 now to see whether they're also included an override, but I doubt it since they appear to be doing the same as CentOS (with v7): run as root, do not drop privileges.

**Comment by deoren** [comment]
I checked both Debian 8 and 9 (matching the same behavior as CentOS 7) and found that both releases just the let the permissions for `/var/log` remain as-is:

`root@debian8:/home/debian# grep '/var/log' /usr/lib/tmpfiles.d/*`
<pre>
/usr/lib/tmpfiles.d/systemd.conf:z /var/log/journal 2755 root systemd-journal - -
/usr/lib/tmpfiles.d/systemd.conf:z /var/log/journal/%m 2755 root systemd-journal - -
/usr/lib/tmpfiles.d/var.conf:d /var/log 0755 - - -
/usr/lib/tmpfiles.d/var.conf:f /var/log/wtmp 0664 root utmp -
/usr/lib/tmpfiles.d/var.conf:f /var/log/btmp 0600 root utmp -
</pre>

Ubuntu appears to have backported the change intended for 16.04 as far back at 15.04 as I found the `/usr/lib/tmpfiles.d/00rsyslog.conf` file there too.

`root@ubuntu-virtual-machine:~# lsb_release -a`
<pre>
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 15.04
Release:	15.04
Codename:	vivid
</pre>

`root@ubuntu-virtual-machine:~# cat /usr/lib/tmpfiles.d/00rsyslog.conf `
<pre>
# Override systemd's default tmpfiles.d/var.conf to make /var/log writable by
# the syslog group, so that rsyslog can run as user.
# See tmpfiles.d(5) for details.

# Type Path    Mode UID  GID  Age Argument
d /var/log 0775 root syslog -
</pre>

**Comment by davidelang** [comment]
dropping privileges and creating files that are secure are problems to make work 
together. Unless you give the user you are switching to the permission to create 
files in the directory, you will have problems.

> From what I can tell that functionality was introduced in their package [in bug 1539483:](https://launchpad.net/bugs/1539483). I found this in their changelog:
>
> <pre>
> rsyslog (8.16.0-1ubuntu1) xenial; urgency=low
>
>  * Merge from Debian unstable (LP: #1539483).  Remaining changes:
>    - debian/00rsyslog.conf Install tmpfiles.d snippet to ensure that the syslog
>      group can write into /var/log/.
> </pre>
>
> I'm checking Debian 8 and 9 now to see whether they're also included an 
> override, but I doubt it since they appear to be doing the same as CentOS 
> (with v7): run as root, do not drop privileges.

So the problem is that Ubuntu is changing things in ways that we don't know 
about, and the complaint is that we aren't doing it the way they are...

**Comment by deoren** [comment]
> So the problem is that Ubuntu is changing things in ways that we don't know
about, and the complaint is that we aren't doing it the way they are...

Well, I'm not sure. Let me ask this: Is the intention to have rsyslog able to create dynamic files within `/var/log` without the user needing to create an override for `/var/log` to allow that behavior? Assuming that it is, everything else that I've said applies. If the intention is not there, much of what I've said and have to say below does not apply.

Some observations re `/var/log` ownership/permissions and Ubuntu LTS releases:

* Ubuntu 12.04 appears to have `/var/log` owned by `root:root`, set set to 755
* Ubuntu 14.04 has it owned by `root:syslog`, set to 775
* Ubuntu 16.04 has it owned by `root:syslog`, set to 755 via default systemd configuration file (`/usr/lib/tmpfiles.d/var.conf`) and then overridden to 775 by `/usr/lib/tmpfiles.d/20rsyslog.conf` 

Now that systemd is part of the picture and `/var/log` is being set to 0755 by default, group write permissions are no more, unless overridden.

My report was a request that the Ubuntu PPA package provide the same override as the Ubuntu stock package, provided that the development team feels that behavior between the two packages (the original and the upgraded replacement) should be 1:1.

As a user who was not even aware that systemd managed the creation of files, permissions and other settings I was horribly confused by my inability to change permissions on `/var/log` and have them persist after a reboot. This ignorance was brought to the fore when I attempted to use dynamic file creation within the root of `/var/log` and nothing happened.

Thankfully rsyslog did log the permission failure, but it took forever for me to figure out that it was systemd that was resetting my changes. Only after many hours of research did I manage to confirm that and learn both that Ubuntu provides an override to the default systemd (enforced) permissions on `/var/log` and that local administrators can do the same by creating a conf file within `/etc/tmpfiles.d`.

Now that I know what the problem is I can fix it locally, but I believe that if the intention is to allow dynamic file creation within `/var/log`, an override for systemd-enabled systems should be provided by the package to set the needed ownership/permissions to match that distro release.

I hope I'm not coming across as rude/demanding, I'm just trying to point out what I feel is an omission in the support provided by the package for using the PPA package as a 1:1 replacement for the stock Ubuntu package (at least as far as 16.04+ is concerned).

**Comment by davidelang** [comment]
Rsyslog expects to be able to create files in /var/log, it appears that 
Ubuntu/systemd have made this harder to do in recent versions. I makes no sense 
to me why they would be doing this.

it's important to remember that the PPA packages are not specific to different 
versions of Ubuntu, so when the Ubuntu behavior changes from version to version, 
it's impossible to be a 1:1 replacement for everything.

which version of Ubuntu should the PPA packages match? the most current version? 
the most recent LTS version?, the LTS version that most people are using (and 
how do you tell which that is)?

This is not the first time that systemd has broken other software, and you 
really should blame systemd for such breakage, not the other software.

In this case, the /usr/lib/tempfile.d and any config files it contains are 
specific to the Ubuntu package. If it was to call these files config files, then 
when that package is uninstalled in favor of the adiscon PPA packages, these 
files would not be deleted. By not calling them config files, the Ubuntu 
packages are arguably broken because if you edit those files, the Ubuntu 
packages will undo your edits.

If systemd is going out and changing system permissions that have been set by 
the system administrator, that's a bug with systemd (or at least the way the 
distro configures systemd)

**Comment by deoren** [comment]
> it's important to remember that the PPA packages are not specific to different
versions of Ubuntu, so when the Ubuntu behavior changes from version to version,
it's impossible to be a 1:1 replacement for everything.

Understood.

> which version of Ubuntu should the PPA packages match? the most current version?
the most recent LTS version?, the LTS version that most people are using (and
how do you tell which that is)?

Looking at the PPA options, it appears that packages for Ubuntu releases from 12.04 to 17.04 are currently available. I suspect that if the PPA packages were to provide the `/usr/lib/tmpfiles.d/00rsyslog.conf` file that versions without systemd support for tmpfiles.d would just ignore the file. I would expect (but haven't confirmed) that Ubuntu 12.04 ESM is likely the last distro without systemd tmpfiles.d support.

Since Ubuntu began providing the override with 15.04 and up, I would say that setting that as a target for including the `/usr/lib/tmpfiles.d/00rsyslog.conf` file makes sense.

Re future changes and supporting them, I guess it would depend on whether future Ubuntu releases break functionality in a way that you can reliably work around when a user installs the rsyslog package provided by the PPA.

I suppose you could also modify the PPA package to display an alert regarding support for dynamically generating log files at the root of `/var/log`. Basically tell the user they should copy the `/usr/share/rsyslog/00rsyslog-varlog.conf` file to `/etc/tmpfiles.d/` if they wish to restore that behavior.

**Comment by rgerhards** [comment]
note that this is probably an issue for the packaging project. maybe @friedl can have a look here. A PR against the packaging project would probably be most welcome (note that I right now just skimmed over the issue,but I thought I provide the info asap).

**Comment by deoren** [comment]
@rgerhards Are you able to move this issue to the packaging project or should I close this one and recreate there?

**Comment by deoren** [comment]
Created rsyslog/rsyslog-pkg-ubuntu#61. Closing this issue in favor of that one.
